feat: add support for JSON data type in Spanner

olavloite opened the pull request:
Introduces support for the new `JSON` data type to Spanner. JSON is represented as plain strings in the client library, and no parsing or checking of the JSON is done by the client. As strings are also used to represent the `STRING` data type in Spanner, there is no default mapping from one Clr type to the JSON data type. That means that a client application __must__ set `SpannerDbType.Json` as the type of a parameter that should be interpreted and sent as JSON to Spanner. (This is equal to the behavior of the `DATE` data type that also does not have a corresponding Clr type, as `DateTime` is already used for `TIMESTAMP`.)

jskeet commented:
I will have a look at this when I get the chance, but it may not be before Tuesday (due to vacation and a public holiday).

skuruppu reviewed (commented):
Looks good from my perspective but I will let @jskeet review first before approving.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/GetSchemaTableTests.cs:
```diff
@@ -45,11 +44,15 @@ public async Task GetSchemaTable_WithFlagEnabled_ReturnsNull()
             using (var connection = new SpannerConnection($"{_fixture.ConnectionString};EnableGetSchemaTable=true"))
             {
-                var command = connection.CreateSelectCommand($"SELECT Int64Value, BytesArrayValue FROM {_fixture.TableName}");
+                var sql = _fixture.RunningOnEmulator
```
I think it would be cleaner to have a separate to for JSON schema values. Most of what's tested is here doesn't really correspond with "GetSchemaTable with the flag enabled returns null".

olavloite replied:
I think that this test case more in general could need some work:
1. The `_ReturnsNull` suffix indicates that the method should return `null` with the flag enabled, but it actually does the opposite: Enabling the flag will ensure that the method does __not__ return null.
2. The current method only checks the return values of `Int64` and `BytesArray` columns. I think it would be valuable to cover all the data types in this test case. It already uses a table that contains columns with all supported data types, so it would only be a matter of adding assertions for the other columns as well. (This is the reason I chose to add the assertions for JSON here)

WDYT?

olavloite replied:
Yes, that shouldn't be a problem, it will probably take a little longer before this PR can be merged.

(I'm also happy to take it on, if that would be easier for you)

jskeet replied:
If you've got the time, that would be great, thanks. Feel free to change your mind and push it back to me :)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerDbTypeTests.ValueConversions.cs:
```diff
@@ -54,13 +54,14 @@ public enum TestType
             { "TimestampField", SpannerDbType.Timestamp, new DateTime(2017, 1, 31, 3, 15, 30) },
-            { "NumericField", SpannerDbType.Numeric, SpannerNumeric.MaxValue }
+            { "NumericField", SpannerDbType.Numeric, SpannerNumeric.MaxValue },
+            { "JsonField", SpannerDbType.Json, "some-json-string" }
```
Is the value of a JSON field meant to be a valid JSON token? If so, I think we should make the test reflect that. Just putting it in double quotes would make it valid for a JSON string token.

olavloite replied:
I've changed it to an empty JSON string (`{}`)

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerDbTypeTests.ValueConversions.cs:
```diff
@@ -113,6 +114,16 @@ private static IEnumerable<SpannerNumeric> GetSpannerNumericsForArray()
 
+        private static IEnumerable<string> GetJsonStringsForArray()
+        {
+            // These are not well-formatted JSON strings, but that is not checked
```
Again, I'd far rather use valid data if possible. (It's not clear what you mean by it preventing problems with tests that serialize struct values to JSON... sure, we might end up with some escaping going on, but I'd like to see that in tests anyway.)

olavloite replied:
I've changed the input and expected values to valid JSON strings (including the necessary escaping).

jskeet reviewed (commented):
Mostly fine - I'd just like to see valid values in the tests.

jskeet reviewed (commented):
Out of interest, do we expect the emulator to support JSON at the same time as the production service? If so, I think I'd prefer to see the PR with unconditional tests that fail on the emulator... the code will be so much simpler.

olavloite commented:
> Out of interest, do we expect the emulator to support JSON at the same time as the production service? If so, I think I'd prefer to see the PR with unconditional tests that fail on the emulator... the code will be so much simpler.

No, unfortunately the emulator will normally implement new features of Cloud Spanner with a delay of 2-3 months (but could be more, could be less). For the `NUMERIC` data type, which was the previous similar addition, it took a little more than 3 months between the official release and before it was available on the emulator.

olavloite commented:
> I think you might be missing an update to BindingTests

Good point, thanks. I've added tests for `JSON` to the `BindingTests` as well.

I've also removed the special handling of `NUMERIC` from the tests here: #6456

skuruppu commented:
@olavloite if you could handle the conflicts on this PR, that would be awesome.

@jskeet and @amanda-tarafa if you could please take another look and approve, that would be great. We're getting close to being able to merge this PR.

jskeet reviewed (approved):
Just re-reviewed - looks fine to me, thanks. Will merge after collisions have been resolved.

jskeet commented:
Sorry, just re-read - I *won't* merge until given the green light explicitly :)

jskeet commented:
https://github.com/googleapis/google-cloud-dotnet/pull/6659 now includes the JSON type - does that mean this can be rebased and merged?

skuruppu commented:
Since the tests are passing, yes absolutely :)

olavloite commented:
> #6659 now includes the JSON type - does that mean this can be rebased and merged?

It's now been rebased and cleaned up. As far as I'm concerned, it is ready to be merged.

jskeet commented:
Hmm... it's showing conflicts at the moment. Could you rebase against master and force push again? I'll then review (which may well be tomorrow rather than today) and merge if I have no further comments.

olavloite commented:
> Hmm... it's showing conflicts at the moment. Could you rebase against master and force push again? I'll then review (which may well be tomorrow rather than today) and merge if I have no further comments.

Huh, I'm not seeing any conflicts at the moment... Could it be that your browser is loading the PR from cache?

jskeet commented:
> Could it be that your browser is loading the PR from cache?

Nope, force reloaded. This is what I'm seeing at the bottom of the PR - which may be because I've got the "Rebase and merge" option selects (and that's what I'd want to use):

![Rebase failure](https://user-images.githubusercontent.com/17011/123240914-a3bc8e80-d4d8-11eb-8f86-8db5f8816979.png)

olavloite commented:
Ah, I see. (The `Squash and merge` which is the default I'm seeing does not indicate any merge problems). I'll have a look.

olavloite commented:
I've rebased and squashed it down to one commit. PTAL when you have time.

jskeet commented:
Great, thanks - have been advised to hold off actually merging temporarily anyway, but I'll review when I can just to remind myself of it, and be ready to merge when given the go-ahead.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/AllTypesTableFixture.cs:
```diff
@@ -30,25 +30,53 @@ public AllTypesTableFixture() : base("TypesTable")
-                            TimestampArrayValue ARRAY<TIMESTAMP>,
-                            DateArrayValue      ARRAY<DATE>,
-                        ) PRIMARY KEY(K)");
+            // The emulator doesn't yet support the NUMERIC type.
```
This comment should refer to JSON rather than NUMERIC, right?

olavloite replied:
Yes, that change seems to have been lost in the rebase. Fixed now.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/AllTypesTableFixture.cs:
```diff
@@ -30,25 +30,53 @@ public AllTypesTableFixture() : base("TypesTable")
+                              Int64ArrayValue     ARRAY<INT64>,
+                              Float64ArrayValue   ARRAY<FLOAT64>,
+                              NumericArrayValue   ARRAY<NUMERIC>,
+                              StringArrayValue    ARRAY < STRING(MAX) >,
```
Are the spaces in these types necessary, or would `ARRAY<STRING(MAX)>` work?

olavloite replied:
The spaces are not needed, so removed.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/GetSchemaTableTests.cs:
```diff
@@ -101,8 +104,9 @@ public async Task GetSchemaTable_WithFlagEnabled_ReturnsColumnOrdinals()
-                    for (var ordinal = 1; ordinal <= SchemaTestData.Count(); ordinal++)
+                    var expectedRowCount = _fixture.RunningOnEmulator ? SchemaTestData.Count() - 1 : SchemaTestData.Count() + 1;
+                    Assert.Equal(expectedRowCount, table.Rows.Count);
+                    for (var ordinal = 1; ordinal <= expectedRowCount - 1; ordinal++)
```
Possibly `ordinal < expectedRowCount` instead?

olavloite replied:
Done

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/WriteTests.cs:
```diff
@@ -109,6 +110,13 @@ public async Task WriteValues()
                 { "NumericArrayValue", SpannerDbType.ArrayOf(SpannerDbType.Numeric), nArray }
             };
 
+            // The emulator doesn't yet support the NUMERIC and JSON types.
```
Adjust comment to just JSON

olavloite replied:
Done.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:
```diff
@@ -162,6 +162,12 @@ internal Value ToProtobufValue(object value, SpannerConversionOptions options)
+                case TypeCode.Json:
+                    if (value is string jsonString)
+                    {
+                        return new Value { StringValue = jsonString };
```
Note to self: after this is merged, I'll rewrite all of these "new Value" to "Value.FromString" etc.

jskeet reviewed (approved):
Basically still fine - a few comments that could be cleaned up. I'm going to hold off merging as advised even after that though.

skuruppu commented:
If everyone is ok with it, and since the tests are passing, let's merge this PR :)
